scene/resources/material.h

```diff
@@ -235,68 +248,82 @@ class BaseMaterial3D : public Material {
 		SPECULAR_PHONG,
 		SPECULAR_TOON,
 		SPECULAR_DISABLED,
+		SPECULAR_MAX
 	};
 
 	enum BillboardMode {
 		BILLBOARD_DISABLED,
 		BILLBOARD_ENABLED,
 		BILLBOARD_FIXED_Y,
 		BILLBOARD_PARTICLES,
+		BILLBOARD_MAX
 	};
 
 	enum TextureChannel {
 		TEXTURE_CHANNEL_RED,
 		TEXTURE_CHANNEL_GREEN,
 		TEXTURE_CHANNEL_BLUE,
 		TEXTURE_CHANNEL_ALPHA,
-		TEXTURE_CHANNEL_GRAYSCALE
+		TEXTURE_CHANNEL_GRAYSCALE,
+		TEXTURE_CHANNEL_MAX
 	};
 
 	enum EmissionOperator {
 		EMISSION_OP_ADD,
-		EMISSION_OP_MULTIPLY
+		EMISSION_OP_MULTIPLY,
+		EMISSION_OP_MAX
 	};
 
 	enum DistanceFadeMode {
 		DISTANCE_FADE_DISABLED,
 		DISTANCE_FADE_PIXEL_ALPHA,
 		DISTANCE_FADE_PIXEL_DITHER,
 		DISTANCE_FADE_OBJECT_DITHER,
+		DISTANCE_FADE_MAX
 	};
 
 private:
 	union MaterialKey {
+		// parameters values
 		struct {
+			// enums
+			uint64_t texture_filter : TEXTURE_FILTER_MAX;
```

Comment:
This does not make sense, the idea is making sure the bits cover the amount of values in the enumeration, using max increases the amount of bits significantly.

Reply:
It was still technically bounded by the two unit64_t's in the union, but I have updated them to be MAX - 1.

Also, I converted MaterialKey to a struct and used memsets and memcmp's so that its simpler.